module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,72 @@
+#include "advancedCalculator.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <map>
+#include <regex>
+
+const std::map<std::string, std::function<double(double, double)>> commands{
+    {"+", std::plus<double>()},
+    {"-", std::minus<double>()},
+    {"*", std::multiplies<double>()},
+    {"/", std::divides<double>()},
+    {"%", std::modulus<int>()},
+    {"^", [](double base, double exponent) { return pow(base, exponent); }},
+    {"$", [](double num, double root) { return pow(num, 1.0 / root); }}
+};
+
+void exportKeys(std::string& operators) {
+    for (const auto& el : commands) {
+        operators.append(el.first);
+    }
+}
+
+bool checkCharacters(const std::string& input) {
+    std::string allowedSigns{" ,.!"};
+    exportKeys(allowedSigns);
+
+    return std::any_of(input.cbegin(), input.cend(), [&allowedSigns](char letter) {
+        return std::find(allowedSigns.cbegin(), allowedSigns.cend(), letter) == allowedSigns.cend() && !isdigit(letter);
+    });
+}
+
+bool isNumFloat(const std::string& num) {
+    return stoi(num) != stod(num);
+}
+
+double factorial(double num) {
+    if (num <= 0.0) {
+        return 1.0;
+    }
+
+    return tgamma(num + 1.0);
+}
+
+ErrorCode process(std::string input, double* out) {
+    if (checkCharacters(input)) {
+        return ErrorCode::BadCharacter;
+    }
+
+    const std::regex commandRegex(R"((([-]?\d+[.]*\d*)[ ]*([+\-*\/$^%])[ ]*([-]?\d+[.]*\d*))|(([-]?\d+[.]*\d*)[ ]*!))");
```

Comment:
My suggestion:

Let's split these regex. One for "operations" case, and second for "factorial operation" case. Inside factorial operation you can define variable defaultValue as const variable. In my opinion it will make your code clearer to read :p. It's just a suggestion ;~}. You can check it in my pr if you want to :)

Reply:
+1